cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -562,6 +563,33 @@ Result<std::shared_ptr<Schema>> ExtractSchemaToBind(const compute::Declaration&
   } else if (declr.factory_name == "filter") {
     auto input_declr = std::get<compute::Declaration>(declr.inputs[0]);
     ARROW_ASSIGN_OR_RAISE(bind_schema, ExtractSchemaToBind(input_declr));
+  } else if (declr.factory_name == "project") {
+    auto input_declr = std::get<compute::Declaration>(declr.inputs[0]);
+    ARROW_ASSIGN_OR_RAISE(auto input_schema, ExtractSchemaToBind(input_declr));
+    const int num_fields_before_proj = input_schema->num_fields();
+    const auto& opts = checked_cast<const compute::ProjectNodeOptions&>(*(declr.options));
+    const auto& exprs = opts.expressions;
+    int i = 0;
+    bind_schema = input_schema;
+    for (const auto& expr : exprs) {
```

Comment:
Minor nit: I have a slight preference for:

```
for (int i = 0; i < static_cast<int>(exprs.size()); i++) {
  const auto& expr = exprs[i];
  // ...
}
```

over

```
int i = 0;
for (const auto& expr : exprs) {
  // ...
  i++;
}
```
  
In other words, prefer range based for loop if you can but don't force it if you can't.  That being said, this preference may just be me (I don't think the google style guide mandates it) so feel free to ignore this suggestion.

Reply:
Let's go with your style, it has more advantages when we have to use the index.